sdk/python/lib/pulumi/runtime/rpc.py

```diff
@@ -208,23 +257,50 @@ async def serialize_property(value: 'Input[Any]',
             }
         return value
 
-    transform_keys = True
-
-    # If value is an input type, convert it to a dict, and set transform_keys to False to prevent
-    # transforming the keys of the resulting dict as the keys should already be the final names.
+    # If value is an input type, convert it to a dict.
     value_cls = type(value)
     if _types.is_input_type(value_cls):
         value = _types.input_type_to_dict(value)
-        transform_keys = False
+        types = _types.input_type_types(value_cls)
+
+        return {
+            k: await serialize_property(v, deps, input_transformer, types.get(k))
+            for k, v in value.items()
```

Comment:
Is this likely to run into issues if the input type has an `items` property?

I think no because `value` is already a `dict`, but just making sure.

Reply:
> I think no because `value` is already a `dict`, but just making sure.

Right. It's guaranteed to be a `dict` here, so we don't have to worry about that.